### PR TITLE
Fix cmake toolchain compile flags loop

### PR DESCRIPTION
**Pull request by BredPet**
- fix https://github.com/dotnet/corert/issues/5093
	- cmake toolchain refactoring

### CODE REVIEW

**Comment by BredPet** [comment]
@dotnet/arm32-contrib please review

File: cross/arm/toolchain.cmake [hunk 1]
```diff
@@ -1,38 +1,2 @@
...
-    set(WITH_LLDB_LIBS "${LLVM_ARM_DIR}/lib/" CACHE STRING "")
-    set(WITH_LLDB_INCLUDES "${LLVM_ARM_DIR}/include" CACHE STRING "")
-    set(LLDB_H "${WITH_LLDB_INCLUDES}" CACHE STRING "")
-    set(LLDB "${LLVM_ARM_DIR}/lib/liblldb.so" CACHE STRING "")
-else()
-    set(WITH_LLDB_LIBS "${CROSS_ROOTFS}/usr/lib/${TOOLCHAIN}" CACHE STRING "")
-    set(WITH_LLDB_INCLUDES "${CROSS_ROOTFS}/usr/lib/llvm-3.6/include" CACHE STRING "")
-endif()
+get_filename_component(ARCH_NAME "${CMAKE_CURRENT_LIST_DIR}" NAME)
+include(${CMAKE_CURRENT_LIST_DIR}/../crossoptions.cmake)
```
Comment (kasper3) [inline cross/arm/toolchain.cmake]: If all these files have identical two lines left, should they be removed and callsites be updated instead?

**Comment by janvorli** [comment]
@BredPet I am not sure why the title of the PR is "fix compile flags loop". What were you referring to as "loops"?

**Comment by BredPet** [comment]
https://github.com/dotnet/corert/pull/5091#discussion_r156154156
For example: without changes, if you build x86 arch with x86 rootfs, after 10 attempts you'll receive something like this:
```
CMAKE_EXE_LINKER_FLAGS =   --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32
CMAKE_SHARED_LINKER_FLAGS =   --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32
CMAKE_MODULE_LINKER_FLAGS =   --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32  --sysroot=/home/user/work/coreclr/cross/rootfs/x86 -B/home/user/work/coreclr/cross/rootfs/x86/usr/lib/gcc/i686-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/lib/i386-linux-gnu -L/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -Wl,-rpath-link,/home/user/work/coreclr/cross/rootfs/x86/usr/lib/i386-linux-gnu -m32
```
and so on.
After 30 attempts or more, you will get a compilation error due to overflow of the compilation flags.

File: cross/tryrun.cmake [hunk 2]
```diff
@@ -0,0 +1,45 @@
+set(TARGET_ARCH_NAME $ENV{TARGET_BUILD_ARCH})
+
+macro(set_cache_value)
+  set(${ARGV0} ${ARGV1} CACHE STRING "Result from TRY_RUN" FORCE)
+endmacro()
+
+if(TARGET_ARCH_NAME MATCHES "^(armel|arm|arm64|x86)$")
+  set_cache_value(FILE_OPS_CHECK_FERROR_OF_PREVIOUS_CALL_EXITCODE 1)
```
[the remark this answers is not in the document]
Comment (BredPet) [reply cross/tryrun.cmake]: Just moved it as it was before for each architecture separately.
I chose their common matching part.

File: cross/toolchain.cmake [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+elseif(TARGET_ARCH_NAME STREQUAL "arm64")
+  include_directories(SYSTEM ${CROSS_ROOTFS}/usr/include)
+endif()
+
+# add_compile_param - adds only new options without duplicates.
+# arg0 - list with result options, arg1 - list with new options.
+# arg2 - quick summary string for optional using CACHE FORCE mode.
+macro(add_compile_param)
+  if(NOT ${ARGC} MATCHES "^(2|3)$")
+    message(FATAL_ERROR "Wrong using add_compile_param!")
```
[the remark this answers is not in the document]
Comment (BredPet) [reply cross/toolchain.cmake]: Ok, I'll add a few details.

**Pull request merged**